Pinta.Effects/Effects/AlignObjectEffect.cs

```diff
@@ -0,0 +1,169 @@
+	{
+		// If no selection, it's the whole image
+		RectangleI selection = rois[0];
+		AlignPosition align = Data.Position;
+
+		RectangleI objectBounds = Utility.GetObjectBounds (src, selection);
+
+		// Calculate the new position for the object
+		PointI newPosition = CalculateNewPosition (objectBounds, align, selection);
+
+		// Draw the object in the new position
+		MoveObject (src, dest, objectBounds, newPosition, selection);
+	}
+
+	private PointI CalculateNewPosition (RectangleI objectBounds, AlignPosition align, RectangleI selectionBounds)
+	{
+		int x = 0;
+		int y = 0;
+
+		// Align with the selection bounds
+		switch (align) {
+			case AlignPosition.TopLeft:
+				x = selectionBounds.X;
+				y = selectionBounds.Y;
+				break;
+			case AlignPosition.TopCenter:
+				x = selectionBounds.X + selectionBounds.Width / 2 - objectBounds.Width / 2;
+				y = selectionBounds.Y;
+				break;
+			case AlignPosition.TopRight:
+				x = selectionBounds.Right - objectBounds.Width;
+				y = selectionBounds.Y;
+				break;
+			case AlignPosition.CenterLeft:
+				x = selectionBounds.X;
+				y = selectionBounds.Y + selectionBounds.Height / 2 - objectBounds.Height / 2;
+				break;
+			case AlignPosition.Center:
+				x = selectionBounds.X + selectionBounds.Width / 2 - objectBounds.Width / 2;
+				y = selectionBounds.Y + selectionBounds.Height / 2 - objectBounds.Height / 2;
+				break;
+			case AlignPosition.CenterRight:
+				x = selectionBounds.Right - objectBounds.Width;
+				y = selectionBounds.Y + selectionBounds.Height / 2 - objectBounds.Height / 2;
+				break;
+			case AlignPosition.BottomLeft:
+				x = selectionBounds.X;
+				y = selectionBounds.Bottom - objectBounds.Height;
+				break;
+			case AlignPosition.BottomCenter:
+				x = selectionBounds.X + selectionBounds.Width / 2 - objectBounds.Width / 2;
+				y = selectionBounds.Bottom - objectBounds.Height;
+				break;
+			case AlignPosition.BottomRight:
+				x = selectionBounds.Right - objectBounds.Width;
+				y = selectionBounds.Bottom - objectBounds.Height;
+				break;
+		}
+
+		return new PointI (x, y);
+	}
+
+	private void MoveObject (ImageSurface src, ImageSurface dest, RectangleI objectBounds, PointI newPosition, RectangleI selectionBounds)
+	{
+		var src_data = src.GetReadOnlyPixelData ();
+		var dst_data = dest.GetPixelData ();
+		int width = src.Width;
+
+		// Clear the selection area
+		for (int y = 0; y < selectionBounds.Height; y++) {
+			var dst_row = dst_data.Slice ((selectionBounds.Y + y) * width + selectionBounds.X, selectionBounds.Width);
+
+			for (int i = 0; i < dst_row.Length; i++) {
+				dst_row[i] = src.GetColorBgra (PointI.Zero);
+			}
+		}
+
+		// Draw the object in the new position
+		for (int y = 0; y < objectBounds.Height; y++) {
+			var src_row = src_data.Slice ((objectBounds.Y + y) * width + objectBounds.X, objectBounds.Width);
+			var dst_row = dst_data.Slice ((newPosition.Y + y) * width + newPosition.X, objectBounds.Width);
+
+			for (int i = 0; i < src_row.Length; i++) {
```

Comment:
Would `src_row.CopyTo(dst_row)` work here instead of the loop?

Reply:
i knew there *had* to be a better way, but i couldn't find it :sweat_smile: 
it seems to work, thanks for the tip!